File: example/RTC4/DEMO3.cs
//  File: DEMO3.cs
//----------------------------------------------------------------------------
//  Copyright (c) 2015 by SCANLAB AG.                     All rights reserved.
//----------------------------------------------------------------------------
//
//
//  Abstract
//      A console application for continuously plotting Archimedean spirals
//      by means of a Nd:YAG laser
//
//  Author
//      Bernhard Schrems, SCANLAB AG
//
//  Features
//      - explicit linking to the RTC4DLL.DLL or RTC4DLLx64.DLL
//      - use of both list buffers for continuous data transfer
//      - exception handling
//
//  Comment
//      This routine demonstrates how to use both list buffers for a
//      continuous data transfer by applying the command "auto_change"
//      on a loaded list buffer. Methods to halt and to resume the data
//      transfer are also shown.
//      Notice that the spirals are only exposing as Archimedean spirals,
//      if there is a F-Theta-lens in use.
//
//  Necessary Sources
//      RTC4Wrap.cs
//
//  NOTE
//      THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
//      KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//      IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR
//      PURPOSE.
//
//----------------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using RTC4Import;


namespace Demo
{
    // End Locus of the vector to plot
    struct locus
    {
        public short xval, yval;
        public locus(short x, short y)
        {
            xval = x;
            yval = y;
        }
    }

    class RTC4Exception : ApplicationException
    {
        public RTC4Exception(string ErrMsg) : base(ErrMsg) {}
    }

    class Plotter : RTC4Wrap
    {
        // Desired Nd:YAG Laser Timing for YAG mode 1
        const uint LASER = 10000;           // Laser pulse frequency in Hz.
        const uint PULSE = 5;               // Pulse width in microseconds.
        const uint PULSE1ST = 2;            // Amount of pulses the first
                                            // pulse killer FPK
                                            // should be active.
        // Frequency Base
        const uint MEGA = (1000 * 1000);
        const uint FREQBASE = (8 * MEGA);   // Corresponding to a time base
                                            // of 1/8 microseconds.
        // Derived Laser Timing in microseconds
        const uint FPK = ((100 * PULSE1ST + 50) * FREQBASE / 100 / LASER);
        const uint HALF_PERIOD = (FREQBASE / LASER / 2);
        const uint PULSE_WIDTH = (PULSE * FREQBASE / MEGA);

        uint list = 1;
        uint ListLevel = 0;
        bool ListStart = false;
        bool FirstTime = false;
        bool Suspended = false;

        public Plotter()
        {
            // Initialize the RTC
            short ErrorCode = load_correction_file(
                                    "cor_1to1.ctb",
                                    1,          // table; #1 is used by default
                                    1.0, 1.0,   // scale factor
                                    0.0,        // rotation in degrees, counterclockwise
                                    0.0, 0.0);  // offset in bits
            if (ErrorCode != 0)
            {
                string Msg = String.Format("Correction file loading error: {0}\n\n", ErrorCode);
                throw new RTC4Exception(Msg);
            }
            ErrorCode = load_program_file("RTC4d2.hex");
            if (ErrorCode != 0)
            {
                string Msg = String.Format("Program file loading error: {0}\n\n", ErrorCode);
                throw new RTC4Exception(Msg);
            }

            set_laser_mode(1);      // YAG mode 1 selected
            set_firstpulse_killer((ushort)FPK);

            // Activate a home jump and specify the beam dump 
            locus BeamDump = new locus(-32000, -32000);
            home_position(BeamDump.xval, BeamDump.yval);

            // Turn on the optical pump source and wait for 2 seconds
            // (The following assumes that signal ANALOG OUT1 of the laser connector
            //  controls the pump source)
            write_da_1(640);

            // Timing, delay and speed preset
            // Transmit the following list commands to the list buffer
            set_start_list(1);
            // Wait for 2 seconds
            long_delay(50000);
            long_delay(50000);
            long_delay(50000);
            long_delay(50000);
            set_laser_timing(
                    (ushort)HALF_PERIOD,    // half of the laser signal period
                    (ushort)PULSE_WIDTH, 2, // pulse widths of signal LASER1
                    1);             // time base; 0 corresponds to 1 microsecond.
            // Otherwise, the time base is 1/8 microseconds.
            set_scanner_delays(25,  // jump delay in 10 microseconds
                            10,     // mark delay in 10 microseconds
                            5);     // polygon delay in 10 microseconds
            set_laser_delays(100,   // laser on delay in microseconds
                            100);   // laser off delay in microseconds
            set_jump_speed(1000.0); // jump speed in bits per milliseconds
            set_mark_speed(250.0);  // marking speed in bits per milliseconds
            set_end_of_list();
            execute_list(1);

            // Wait on "Pump source warming up" is done.
            ushort busy, position;
            do
            {
                get_status(out busy, out position);
            } while (busy != 0);
        }

        public bool Plot(locus destination, ref bool StartLocus, ref bool PlotLaunch)
        {
            ushort busy, position;

            if (PlotLaunch)
            {
                do
                {
                    get_status(out busy, out position);
                } while (busy != 0);
                list = 1;
                ListStart = true;
                ListLevel = 0;
                FirstTime = true;
                PlotLaunch = false;
            }
            else
            {
                // On a list swap, check whether the list "list" is free to be loaded
                if (ListStart)
                {
                    get_status(out busy, out position);
                    if (busy != 0)
                        if (list == 2)
                        {
                            if (position >= 4000) return (true);    // list 2 is busy
                        }
                        else
                        {
                            if (position < 4000) return (true);     // list 1 is busy
                        }
                }
            }

            // Open the list at the beginning
            if (ListStart) set_start_list((ushort)list);
            ListStart = false;
            ListLevel++;

            if (StartLocus)
            {
                jump_abs(destination.xval, destination.yval);
                StartLocus = false;
            }
            else
                mark_abs(destination.xval, destination.yval);

            if (ListLevel < 4000 - 1) return (false);   // Done

            set_end_of_list();
            ListStart = true;

            // Only, apply execute_list for the very first time.
            // Otherwise, use auto_change.
            if (FirstTime)
            {
                execute_list(1);
                FirstTime = false;
            }
            else auto_change();

            list = (list == 1) ? (uint)2 : (uint)1;
            ListLevel = 0;

            return (false);                             // Done
        }
        public void Resume()
        {
            if (Suspended)
            {
                restart_list();
                Suspended = false;
            }
            enable_laser();
        }
        public void Suspend()
        {
            stop_list();
            // Subsequent list commands will not be executed
            // as long "stop_list" is active
            Suspended = true;
        }
        public bool IsSuspended()
        {
            return(Suspended);
        }
        public void Flush()
        {
            if (ListStart) return;
            else
            {
                set_end_of_list();
                if (FirstTime)
                {
                    execute_list(1);
                }
                else
                {
                    auto_change();
                }
            }
        }
        public void Stop()
        {
            disable_laser();

            // Remove a pending "stop_list" call before calling 
            // "stop_execution"
            if (Suspended) restart_list();

            stop_execution();

            // Finish
            // Activate the pump source standby
            write_da_1(0);
        }
    }

    class DEMO3 
    {
        static void Main(string[] args)
        {
            // Spiral Parameters
            const double AMPLITUDE = 10000.0;
            const double PERIOD    = 512.0;      // amount of vectors per turn
            const double OMEGA = (2 * Math.PI / PERIOD);

            Console.Title = "RTC®4 Demo Program 3";
            Console.WriteLine("Archimedean spirals\n\n");

            try
            {
                // Initialize the scan-system and the laser.
                Console.WriteLine("Pump source warming up - please wait ...\n");
                Plotter plotter = new Plotter();

                // Plotting Routine
                Console.WriteLine("Press 1, 2, ... or 9 to set the number of revolutions.");
                Console.WriteLine("Press S to suspend or R to resume plotting.");
                Console.WriteLine("Press O to turn off or R to restart plotting.");
                Console.WriteLine("Press F to flush the list buffers and to terminate.");
                Console.WriteLine("Any other key will halt plotting and terminate.\n");

                double turns = 5.0;
                Console.WriteLine("--- revolutions: {0} ---", (int)turns);
                double increment = AMPLITUDE / turns / PERIOD;
                int limit = (int)PERIOD * (int)turns;

                bool IsEventOff = false, flush = false;
                bool StartLocus = true, PlotLaunch = true;
                bool status = false;
                int i = 0;

                locus point = new locus();
                double span = increment;
                for (bool CarryOn = true; CarryOn; i++, IsEventOff = false, span += increment)
                {
                    if (i == limit)
                    {
                        i = 0;
                        span = increment;
                        StartLocus = true;
                    }

                    point.xval = (short)(span * Math.Sin(OMEGA * (double)i));
                    point.yval = (short)(span * Math.Cos(OMEGA * (double)i));

                    while (IsEventOff ? status : plotter.Plot(point, ref StartLocus, ref PlotLaunch))
                    {
                        if (Console.KeyAvailable)
                        {
                            ConsoleKeyInfo ch = Console.ReadKey(true);
                            char LowerCaseKey = char.ToLower(ch.KeyChar);
                            switch (LowerCaseKey)
                            {
                                case '1':
                                    turns = 1.0;
                                    goto TurnsSetup;
                                case '2':
                                    turns = 2.0;
                                    goto TurnsSetup;
                                case '3':
                                    turns = 3.0;
                                    goto TurnsSetup;
                                case '4':
                                    turns = 4.0;
                                    goto TurnsSetup;
                                case '5':
                                    turns = 5.0;
                                    goto TurnsSetup;
                                case '6':
                                    turns = 6.0;
                                    goto TurnsSetup;
                                case '7':
                                    turns = 7.0;
                                    goto TurnsSetup;
                                case '8':
                                    turns = 8.0;
                                    goto TurnsSetup;
                                case '9':
                                    turns = 9.0;
                                TurnsSetup:
                                    // Turns setup requested
                                    if (!IsEventOff && !plotter.IsSuspended())
                                        Console.WriteLine("--- revolutions: {0} ---", (int)turns);
                                    // NOTE
                                    //  Modifying the number of turns might not immediately
                                    //  take effect, because both list buffers can still
                                    //  hold vectors from a previous setting.
                                    increment = AMPLITUDE / turns / PERIOD;
                                    span = 0;
                                    limit = (int)PERIOD * (int)turns;
                                    IsEventOff = true;
                                    StartLocus = true;
                                    i = -1;
                                    break;
                                case 'f':
                                    // Buffer flushing requested
                                    if (!IsEventOff)
                                    {
                                        plotter.Resume();
                                        Console.WriteLine("- flushing the queue -");
                                        flush = true;
                                        IsEventOff = true;
                                        status = false;
                                        CarryOn = false;
                                    }
                                    break;
                                case 's':
                                    // Sudden suspending requested
                                    if (!IsEventOff)
                                    {
                                        plotter.Suspend();
                                        Console.WriteLine("- plotting suspended -");
                                    }
                                    break;
                                case 'r':
                                    // Resume to plot
                                    if (IsEventOff)
                                    {
                                        status = false;
                                        span = 0;
                                        i = -1;
                                        PlotLaunch = true;
                                        StartLocus = true;
                                    }
                                    else
                                    {
                                        plotter.Resume();
                                    }
                                    Console.WriteLine("--- revolutions: {0} ---", (int)turns);
                                    break;
                                case 'o':
                                    // Stop request
                                    plotter.Stop();
                                    Console.WriteLine("- plotting turned off ");

                                    // Do not transfer list commands as long
                                    // "stop_execution" is active
                                    IsEventOff = true;
                                    status = true;
                                    break;
                                default:
                                    plotter.Stop();
                                    Console.WriteLine("-- plotting terminated --\n");
                                    status = false;
                                    CarryOn = false;
                                    IsEventOff = true;
                                    break;
                            }
                        }
                    }
                }

                // Flush the list buffers, on request.
                if (flush) plotter.Flush();
            }
            catch (TypeInitializationException x)
            {
                Console.WriteLine("\n" + x.Message.ToString());
                Console.WriteLine("\t\"" + x.InnerException.Message.ToString() + "\"");
            }
            catch (ApplicationException x)
            {
                Console.WriteLine("Exception: " + x.Message.ToString());
            }
            catch (SystemException x)
            {
                Console.WriteLine("System Exception: " + x.Message.ToString());
            }
            Console.WriteLine("\nFinished - press any key to exit ");
            while (!Console.KeyAvailable) ;
            Console.ReadKey();
       }
    }
}

File: example/RTC4/DEMO3.cpp
//  File
//      DEMO3.cpp
//
//  Abstract
//      A console application for continuously plotting Archimedean spirals
//      by means of a Nd:YAG laser
//
//  Author
//      Bernhard Schrems, SCANLAB AG
//
//  Features
//      - explicit linking to the RTC4DLL.DLL
//      - use of both list buffers for continuous data transfer
//      - exception handling
//
//  Comment
//      This routine demonstrates how to use both list buffers for a
//      continuous data transfer by applying the command "auto_change"
//      on a loaded list buffer. Methods to halt and to resume the data
//      transfer are also shown.
//      The spirals are only exposing as Archimedean spirals, if the
//      scan head is built up with a F-Theta-lens.
//
//  Necessary Sources
//      RTC4expl.H, RTC4expl.CPP, MESSAGE.CPP
//
//  Environment: Win32
//
//  Compiler
//      - compiler independent
//      - tested with Visual C++ 5.0 and Visual C++ 6.0


// system header files
#include <windows.h>
#include <stdio.h>
#include <conio.h>
#include <math.h>

// RTC4 header file for explicitly linking to the RTC4DLL.DLL
#include "RTC4expl.h"


// Definition of "pi"
#define     SL_PI       3.14159265358979323846

// Spiral Parameters
#define     AMPLITUDE   10000.0
#define     PERIOD      512.0       // amount of vectors per turn
#define     OMEGA       (2*SL_PI/PERIOD)


// Desired Nd:YAG Laser Timing for YAG mode 1
#define     LASER       10000       // Laser pulse frequency in Hz
#define     PULSE       5           // pulse width in microseconds
#define     PULSE1ST    2           // amount of pulses the first
                                    // pulse killer FPK
                                    // should be active
// Frequency Base
#define     MEGA        (1000*1000)
#define     FREQBASE    (8*MEGA)            // corresponding to a time base
                                            // of 1/8 microseconds
// Derived Laser Timing in microseconds
#define     FPK         ((100*PULSE1ST+50)*FREQBASE/100/LASER)
#define     HALF_PERIOD (FREQBASE/LASER/2)
#define     PULSE_WIDTH (PULSE*FREQBASE/MEGA)


// End Locus of a Vector
struct locus {
    short xval, yval;
};

// Beam Dump Location
locus BeamDump = {
    -32000, -32000
};


void ErrorMessage(short ErrorCode);
int PlotVector(const locus &destination, unsigned *start, unsigned *PlotLaunch);
void PlotFlush(void);


void main(void *, void *)
{
    unsigned short  busy, position;
    short           ErrorCode;
    unsigned        CarryOn, stopped, start, eventOff, PlotLaunch, flush;
    int             i, limit, status;
    locus           point;
    double          turns, span, increment;

	printf("Archimedean spirals\n\n");

    // Initialize the RTC4DLL.DLL
    if(RTC4open()) {
        printf("Error: RTC4DLL.DLL not found\n");
        printf("Press any key to shut down\n");
        while(!kbhit()) ;
        (void)getch();
        return;
    }


    // Initialize the RTC

    ErrorCode = load_correction_file("cor_1to1.ctb",
                        1,          // table; #1 is used by default
                        1.0, 1.0,   // scale factor
                        0.0,        // rotation in degrees, counterclockwise
                        0.0, 0.0);  // offset in bits
    if(ErrorCode) {
        printf("Correction file loading error: ");
        ErrorMessage(ErrorCode);
        return;
    }
    ErrorCode = load_program_file("RTC4D2.hex");
    if(ErrorCode) {
        printf("Program file loading error: ");
        ErrorMessage(ErrorCode);
        return;
    }
    set_laser_mode(1);      // YAG mode 1 selected
    set_firstpulse_killer(FPK);

    // Activate a home jump and specify the beam dump 
    home_position(BeamDump.xval, BeamDump.yval);

    // Turn on the optical pump source and wait for 2 seconds
    // (The following assumes that signal ANALOG OUT1 of the laser connector
    //  controls the pump source)
    write_da_1(640);

    // Timing, delay and speed preset
    // Transmit the following list commands to the list buffer
    set_start_list(1);
    // Wait for 2 seconds
    long_delay(50000);
    long_delay(50000);
    long_delay(50000);
    long_delay(50000);
    set_laser_timing(
            HALF_PERIOD,    // half of the laser signal period
            PULSE_WIDTH, 2, // pulse widths of signal LASER1
            1);             // time base; 0 corresponds to 1 microsecond.
                            // Otherwise, the time base is 1/8 microseconds.
    set_scanner_delays(25,  // jump delay in 10 microseconds
                    10,     // mark delay in 10 microseconds
                    5);     // polygon delay in 10 microseconds
    set_laser_delays(100,   // laser on delay in microseconds
                    100);   // laser off delay in microseconds
    set_jump_speed(1000.0); // jump speed in bits per milliseconds
    set_mark_speed(250.0);  // marking speed in bits per milliseconds
    set_end_of_list();
    execute_list(1);

    printf("Pump source warming up - please wait\r");
    do {
        get_status(&busy, &position);
    } while(busy);

    // Plot
    printf("Press 1, 2, ... or 9 to set the number of revolutions.\n");
    printf("Press S to suspend or R to resume plotting.\n");
    printf("Press O to turn off or R to restart plotting.\n");
    printf("Press F to flush the list buffers and to terminate.\n");
    printf("Any other key will halt plotting and terminate.\n\n");
    turns = 5.0;
    printf("\r--- revolutions: %d ---", (int)turns);
    increment = AMPLITUDE/turns/PERIOD;
    limit = (int)PERIOD*(int)turns;

    for(flush = status = eventOff = stopped = i = 0, span = increment,
            PlotLaunch = start = CarryOn = 1; CarryOn;
            i++, eventOff = 0, span += increment) {
        if(i == limit) {
            i = 0;
            span = increment;
            start = 1;
        }
        point.xval = (short)(span*sin(OMEGA*(double)i));
        point.yval = (short)(span*cos(OMEGA*(double)i));
        while(eventOff ? status : !PlotVector(point, &start, &PlotLaunch)) {
            if(kbhit()) {
                char ch;

                ch = (char)getch();
                switch(ch) {
                    case '1':
                    case '2':
                    case '3':
                    case '4':
                    case '5':
                    case '6':
                    case '7':
                    case '8':
                    case '9':
                        // Turns setup requested
                        turns = (double)(ch ^ '0');
                        if(!stopped && !eventOff)
                            printf("\r--- revolutions: %d ---", (int)turns);
                        // NOTE
                        //  Modifying the number of turns might not immediately
                        //  take effect, because both list buffers can still
                        //  hold vectors from a previous setting.
                        increment = AMPLITUDE/turns/PERIOD;
                        span = 0;
                        limit = (int)PERIOD*(int)turns;
                        start = eventOff = 1;
                        i = -1;
                        break;
                    case 'f':
                    case 'F':
                        // Buffer flushing requested
                        if(!eventOff) {
                            if(stopped) restart_list();
                            printf("\r- flushing the queue -");
                            flush = eventOff = 1;
                            status = CarryOn = 0;
                        }
                        break;
                    case 's':
                    case 'S':
                        // Sudden suspending requested
                        if(!eventOff) {
                            stop_list();
                            // Subsequent list commands will not be executed
                            // as long "stop_list" is active
                            printf("\r- plotting suspended -");
                            stopped = 1;
                        }
                        break;
                    case 'r':
                    case 'R':
                        // Resume to plot
                        if(eventOff) {
                            status = 0;
                            span = 0;
                            i = -1;
                            PlotLaunch = start = 1;
                        }
                        else {
                            restart_list();
                            stopped = 0;
                        }
                        enable_laser();
                        printf("\r--- revolutions: %d ---", (int)turns);
                        break;
                    case 'o':
                    case 'O':
                        // Stop request
                        // Remove a pending "stop_list" call before calling 
                        // "stop_execution"
                        disable_laser();
                        if(stopped) restart_list();
                        stop_execution();
                        printf("\r- plotting turned off ");
                        // Do not transfer list commands as long
                        // "stop_execution" is active
                        eventOff = status = 1;
                        break;
                    default:
                        // Halt and terminate
                        // Remove a pending "stop_list" call before calling 
                        // "stop_execution"
                        disable_laser();
                        if(stopped) restart_list();
                        stop_execution();
                        printf("\r-- plotting terminated --\n");
                        status = CarryOn = 0;
                        eventOff = 1;
                }
            }
        }
    }
    // Flush the list buffers, on request.
    if(flush)   PlotFlush();

    // Finish
    printf("\nFinished - press any key to terminate ");
    // Activate the pump source standby
    write_da_1(0);
    while(!kbhit()) ;
    (void)getch();
    printf("\n");
    // Close the RTC4.DLL
    RTC4close();
    return;
}



//  PlotVector
//
//  Description:
//
//  Function "PlotVector" tries to transfer the specified vector to one
//  of the two list buffers. If the particular list buffer is filled,
//  "PlotVector" automatically invokes the execution of that list buffer. 
//  A subsequent "PlotVector" call will swap the list buffer and load it
//  with the specified vector.
//  On success, the return is 1.
//  Otherwise, 0 is returned - call function "PlotVector" again till it
//  returns 1. Then "PlotVector" can be called with a next vector.
//  When the vector transfer is finished, call function "PlotFlush" to
//  execute the remaining vectors in the list buffers.
//  The first time "PlotVector" is called with a particular vector,
//  parameter "PlotLaunch" must be set to greater 0. That parameter will
//  automatically be reset to 0 on subsequent "PlotVector" calls.
//
//
//      Parameter   Meaning
//
//      destination The end location of a vector
//
//      start       causes to mark a vector from the current to the
//                  specified location, if "start" is set to 0.
//                  Otherwise, it causes to jump to the specified location
//                  without marking and parameter "start" will automatically
//                  be reset to 0.
//      PlotLaunch  launches the function "PlotVector", if set to greater 0.
//                  That parameter will automatically be reset to 0.

// PlotVector specific variables
unsigned ListStart, FirstTime;

int PlotVector(const locus &destination, unsigned *start, unsigned *PlotLaunch) {
    static unsigned list;
    static unsigned ListLevel;
    unsigned short busy, position;

    if(*PlotLaunch) {
        do {
            get_status(&busy, &position);
        } while(busy);
        list = 1;
        ListStart = 1;
        ListLevel = 0;
        FirstTime = 1;
        *PlotLaunch = 0;
    }
    else {
        // On a list swap, check whether the list "list" is free to be loaded
        if(ListStart) {
            get_status(&busy, &position);
            if(busy)
                if(list == 2) {
                    if(position >= 4000) return(0);     // list 2 is busy
                }
                else {
                    if(position < 4000) return(0);      // list 1 is busy
                }
        }
    }

    // Open the list at the beginning
    if(ListStart)  set_start_list((unsigned short)list);
    ListStart = 0;
    ListLevel++;

    if(*start) {
        jump_abs(destination.xval, destination.yval);
        *start = 0;
    }
    else
        mark_abs(destination.xval, destination.yval);

    if(ListLevel < 4000 - 1)    return(1);      // Success

    set_end_of_list();
    ListStart = 1;

    // Only, apply execute_list for the very first time.
    // Otherwise, use auto_change.
    if(FirstTime) {
        execute_list(1);    
        FirstTime = 0;
    }
    else auto_change();

    list = list == 1 ? 2 : 1;
    ListLevel = 0;

    return(1);                                  // Success
}



//  PlotFlush
//
//  Description:
//
//  The call of that function ensures to empty and to execute the remaining
//  contents of the list buffers.
//

void PlotFlush(void) {
    if(ListStart) return;
    else {
        set_end_of_list();
        FirstTime ? execute_list(1) : auto_change();
    }
}
